lib/FFI/Platypus/Type/Win32/LPCWSTR.pm

```diff
@@ -0,0 +1,123 @@
+package FFI::Platypus::Type::Win32::LPCWSTR;
+
+use strict;
+use warnings;
+use 5.008004;
+use FFI::Platypus;
+use FFI::Platypus::Memory qw( strlenW );
+use Encode qw( decode encode );
+
+# ABSTRACT: Platypus custom type for Windows Unicode strings
+# VERSION
+
+=head1 SYNOPSIS
+
+ use FFI::Platypus;
+
+ my $ffi = FFI::Platypus->new( api => 1 );
+ $ffi->lang('Win32');
+
+ use constant MB_OK => 0;
+
+ $ffi->attach([MessageBoxW => 'MessageBox'] => ['HWND', 'LPCWSTR', 'LPCWSTR', 'UINT'] => 'int');
+
+ MessageBox(undef, "I ❤️ Platypus", "Confession", MB_OK);
+
+=head1 DESCRIPTION
+
+This module provides a L<FFI::Platypus> custom type for Windows Unicode strings.
+The created LPCWSTR will be NULL-terminated.
+
+This type can also be used as a return value. Returned LPCWSTR are expected
+to be NULL-terminated, and they won't be freed.
+
+This type is automatically loaded by C<< $ffi->lang('Win32'); >>.
+
+=cut
+
+use constant _incantation =>
+  $^O eq 'MSWin32' && do { require Config; $Config::Config{archname} =~ /MSWin32-x64/ }
+  ? 'Q'
+  : 'L!';
+
+my @stack;  # To keep buffer alive.
+
+sub perl_to_native
+{
+  if(defined $_[0])
+  {
+    my $buf = encode('UTF-16le', $_[0]."\0");
```

Comment:
Lets compute the wide string encoding instead of hard-coding it.  That way this will work for `wchar_t*` on UNIX as well as `LPCWSTR` on Windows.

I have some example code that can compute the encoding for most already supported platforms in practice, and we can back fill any others that we find not to work later.

https://gist.github.com/plicease/1e87ba3dd3faa8a0bd872b324672839d